File: src/fios-file.c
// SPDX-FileCopyrightText: 2024 Filipe Coelho <falktx@darkglass.com>
// SPDX-License-Identifier: AGPL-3.0-or-later

#include "libfios.h"

#include <stdio.h>
#include <string.h>

#ifdef _WIN32
#include <windows.h>
#else
#include <unistd.h>
#endif

static int usage(char* argv[])
{
    fprintf(stderr, "Usage: %s [r|s] [device-path|auto] [file-path]\n", argv[0]);
    return 1;
}

int main(int argc, char* argv[])
{
    if (argc <= 3)
        return usage(argv);

    bool sending;
    if (!strcmp(argv[1], "r"))
        sending = false;
    else if (!strcmp(argv[1], "s"))
        sending = true;
    else
        return usage(argv);

    const char* devpath;
    if (!strcmp(argv[2], "auto"))
    {
       #if defined(_WIN32)
        devpath = "COM5";
       #elif defined(__aarch64__)
        devpath = "/dev/ttyGS0";
       #else
        devpath = "/dev/ttyACM0";
       #endif
    }
    else
    {
        devpath = argv[2];
    }

    fios_serial_t* const s = fios_serial_open(devpath);

    if (s == NULL)
        return 1;

    fios_file_t* const f = sending ? fios_file_send(s, argv[3])
                                   : fios_file_receive(s, argv[3]);

    if (f == NULL)
    {
        fios_serial_close(s);
        return 1;
    }

    fprintf(stdout, "\n");
    fflush(stdout);

    float progress;
    while (fios_file_idle(f, &progress))
    {
        fprintf(stdout, "\rProgress: %.1f %%", progress * 100);
        fflush(stdout);
       #ifdef _WIN32
        Sleep(100);
       #else
        usleep(100 * 1000);
       #endif
    }

    fprintf(stdout, "\n");
    fflush(stdout);

    fios_file_close(f);
    fios_serial_close(s);
    return 0;
}

File: src/fios-file.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: 2024 Filipe Coelho <falktx@darkglass.com>
# SPDX-License-Identifier: AGPL-3.0-or-later

import sys
import time

from libfios import (
    fios_serial_open,
    fios_serial_close,
    fios_file_send,
    fios_file_receive,
    fios_file_idle,
    fios_file_close,
    fios_file_status_in_progress,
)

def usage():
    print("Usage: %s [r|s] [device-path|auto] [file-path]", sys.argv[0])
    return 1

def main():
    if len(sys.argv) <= 3:
        return usage()

    if sys.argv[1] == "r":
        sending = False
    elif sys.argv[1] == "s":
        sending = True
    else:
        return usage()

    if sys.argv[2] == "auto":
        if sys.platform == 'win32':
            devpath = "COM5"
        else:
            devpath = "/dev/ttyACM0"
    else:
        devpath = sys.argv[2]

    s = fios_serial_open(devpath)

    if not s:
        return 1

    f = fios_file_send(s, sys.argv[3]) if sending else fios_file_receive(s, sys.argv[3])

    if not f:
        fios_serial_close(s)
        return 1

    sys.stdout.write("\n")
    sys.stdout.flush()

    while True:
        r = fios_file_idle(f)
        if r[0] != fios_file_status_in_progress:
            break
        sys.stdout.write("\rProgress: %.1f %%" % (r[1] * 100))
        sys.stdout.flush()
        time.sleep(0.1)

    sys.stdout.write("\n")
    sys.stdout.flush()

    fios_file_close(f)
    fios_serial_close(s)
    return 0

sys.exit(main())
